crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Copyright (c) 2024, Google Inc. */
+
+#ifndef OPENSSL_NO_ML_KEM
+
+# include <assert.h>
+# include <internal/common.h>
+# include <internal/constant_time.h>
+# include <crypto/ml_kem.h>
+# include "ml_kem_local.h"
+
+typedef ossl_ml_kem_cbd_func cbd_t;
+typedef const ossl_ml_kem_vinfo *vinfo_t;
+
+# define DEGREE ML_KEM_DEGREE
+static const int kPrime = ML_KEM_PRIME;
+
+/*
+ * Remainders modulo `kPrime`, for sufficiently small inputs, are computed in
+ * constant time via Barrett reduction, and a final call to reduce_once(),
+ * which reduces inputs that are at most 2*kPrime and is also constant-time.
+ */
+# define BARRETT_SHIFT (2 * ML_KEM_LOG2PRIME)
+static const unsigned kBarrettShift = BARRETT_SHIFT;
```

Comment:
I tried graphing this, it seemed to give a sawtooth function that seemed to always be in range.. (Could have got it wrong I guess)

Reply:
It seems unlikely that Barrett division always gives reduced values for products, you'd need to check all numbers up to $(q-1)^2$.